File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> neighbors;
vector<bool> visited;

int result = 0;

// routes: 현재까지 방문한 노드들
void dfs(vector<int> routes, vector<int>& info) {
    // 현재까지 거쳐온 경로들의 늑대와 양의 수를 센다.
    int lambs = 0;
    int wolves = 0;
    for(int r: routes) {
        if (info[r] == 0) lambs++;
        else wolves++;
    }
    
    // 늑대가 많거나 같으면 양을 모두 잃는다. 탐색을 중단한다.
    if (wolves >= lambs) return;
    
    // 현재까지 구한 양의 개수를 비교해 저장한다.
    result = max(result, lambs);
    
    // 인접노드들 중 방문하지 않는 노드들에 대해 dfs를 수행한다.
    for(int node: routes) {
        // int node = routes[i];
        // 인접 노드들에 대해
        for(int neighbor: neighbors[node]) {
            if (visited[neighbor]) continue;
            
            // 인접 노드를 방문하는 경우의 수에 대해 계산해본다.
            visited[neighbor] = true;
            vector<int> next_routes(routes);
            next_routes.push_back(neighbor);
            // routes.push_back(neighbor);

            dfs(next_routes, info);

            // 인접 노드를 방문하지 않는 경우로 백트래킹한다.
            // routes.pop_back();
            visited[neighbor] = false;
        }
    }
    
}

int solution(vector<int> info, vector<vector<int>> edges) {
    // 가능한 모든 탐색 루트로 탐색해서 양을 가장 많이 데려올 수 있는 경우를 탐색한다.
    int N = info.size();
    
    // 1. 인접 리스트 배열을 초기화한다.
    neighbors.resize(N);
    int src, dst;
    for(vector<int>& edge: edges) {
        src = edge[0];
        dst = edge[1];
        neighbors[src].push_back(dst);
    }
    
    // 2. 방문 배열을 초기화한다.
    visited = vector<bool>(N, false);
    visited[0] = true;
    
    // 3. dfs로 모든 그래프 경로를 탐색하면서 데려올 수 있는 양의 최대 개수를 계산한다.
    vector<int> routes;
    routes.push_back(0);
    dfs(routes, info);
    
    return result;
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    
    static int N;
    static List<Integer>[] neighbors;
    static boolean[] visited;
    
    static int[] infoSt;
    
    static int result = 0;
    
    static void dfs(List<Integer> routes) {
        int lambs = 0, wolves = 0;
        for(int r: routes) {
            if (infoSt[r] == 0) lambs++;
            else wolves++;
        }
        
        // 늑대가 더 많을 경우 양을 모두 잃으므로 탐색 중단
        if (wolves >= lambs) return;
        
        // 양 업데이트
        result = Math.max(result, lambs);
        
        // 이웃 노드들을 방문한다.
        for(int node: routes) {
            
            for(int neighbor: neighbors[node]) {
                if (visited[neighbor]) continue;
                
                // 다음 노드를 방문
                visited[neighbor] = true;
                
                // 방문했을 경우 형성되는 방문 노드 목록
                List<Integer> next = new ArrayList<>(routes);
                next.add(neighbor);
                
                dfs(next); // 추가된 목록을 가지고 dfs
                
                visited[neighbor] = false; // 백트래킹
            }
            
        }
        
    }
    
    public int solution(int[] info, int[][] edges) {
        // edges 길이: N-1
        N = info.length;
        infoSt = info;
        
        // 1. 인접 리스트를 생성한다.
        neighbors = new List[N];
        for(int i=0; i<N; i++) neighbors[i] = new ArrayList<>();
        
        int src, dst;
        for(int[] edge: edges) {
            src = edge[0];
            dst = edge[1];
            neighbors[src].add(dst);
        }
        
        // 2. 방문 배열을 초기화한다.
        visited = new boolean[N];
        Arrays.fill(visited, false);
        visited[0] = true;
        
        // 2. dfs를 돌려 노드를 방문하는 모든 경우의 수를 계산하고, 양이 가장 많은 경우를 카운트한다.
        List<Integer> routes = new ArrayList<>();
        routes.add(0);
        dfs(routes);
        
        return result;
    }
}